_drafts/unittests-cpputest.md

```diff
@@ -0,0 +1,1251 @@
+...
+Linking build/sum/sum_tests
+Running build/sum/sum_tests
+.
+OK (1 tests, 1 ran, 1 checks, 0 ignored, 0 filtered out, 1 ms)
+```
+
+To set CPPUTEST_HOME and TARGET_PLATFORM for your platform, edit the first few
+lines of `MarkefileWorkerOverrides.mk`.
+
+```
+CPPUTEST_HOME ?= /usr/local/Cellar/cpputest/3.8
+TARGET_PLATFORM ?=
+```
+
+## Tips & Tricks
+
+{:.no_toc}
+
+### Debugging a Unit Test
+
+Most unit test frameworks will generate separate binaries for each `.cpp` unit
+test file written so that you can load them in a debugger (lldb or gdb).
+
+```
+$ lldb build/kv_store/kv_store_tests
+(lldb) target create "build/kv_store/kv_store_tests"
+Current executable set to 'build/kv_store/kv_store_tests' (x86_64).
+(lldb) run
+Process 96031 launched 'build/kv_store/kv_store_tests' (x86_64)
+.
+OK (1 tests, 1 ran, 5 checks, 0 ignored, 0 filtered out, 7 ms)
+
+Process 96031 exited with status = 0 (0x00000000)
+```
+
+{:.no_toc}
+
+### Code Coverage
+
+One of the wonderful parts about unit testing is that you can generate a code
+coverage report. This shows which paths were covered in a given set of unit
+tests, so you can be sure that the piece of code was tested in some capacity.
+Note that code coverage doesn't measure the different behaviors a code path
+**could take**, but only that a particular code path **was taken**.
+
+![](img/cpputest-unittests/code_coverage.png)
+
+Above is a an example coverage report taken from the Memfault Public SDK[^5].
+
+{:.no_toc}
+
+### Address Sanitizing
+
+To raise use-after-free and buffer overflow errors in unit tests, use the
+compiler option `-fsanitize=address` when compiling unit tests.
+
+[AddressSanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer)
+
+## C/C++ Unit Tests: Common Issues
+
+Writing unit tests in C isn't as simple as writing tests in some languages. Here
+are some common errors and mistakes that everyone runs into and possible
+solutions.
+
+{:.no_toc}
+
+### Linker Error: Symbol not found
+
+```
+Linking build/kv_store/kv_store_tests
+Undefined symbols for architecture x86_64:
+  "_analytics_inc", referenced from:
+      _kv_store_write in libkv_store.a(kv_store.o)
+      _kv_store_read in libkv_store.a(kv_store.o)
+      _kv_store_delete in libkv_store.a(kv_store.o)
+ld: symbol(s) not found for architecture x86_64
+```
+
+This error is generated by the linker to tell the user that there are undefined
+symbols. In the example above, the function `analytics_inc` is called from three
+different functions, but isn't defined anywhere.
+
+The possible solutions to the issue are:
+
+- Create a fake, stub, or mock file which implements this function and add it
+  either to the headers or compile it into the library
+- Define the function within the unit test file itself, probably at the top of
+  the file. If it is a C function being called by C code, place it within the
+  `extern "C" {}` section.
+- Compile out the call points using a define for unit tests. e.g.
```

Comment:
maybe ... might add "as a last resort" somewhere, not a huge fan of unit test ifdefs